File: src/anylist/recipe_collection.py
"""Recipe collection class for AnyList recipe collections."""

import uuid as uuid_module
from typing import TYPE_CHECKING, Any, Dict, Optional
from typing import List as ListType

if TYPE_CHECKING:
    from .anylist import AnyList


class RecipeCollection:
    """
    Recipe collection class representing a collection of recipes.

    Attributes:
        identifier: Unique identifier for this collection
        name: Collection name
        recipe_ids: List of recipe IDs in this collection
    """

    def __init__(self, collection_data: Dict[str, Any], anylist_instance: "AnyList"):
        self._identifier = collection_data.get("identifier") or str(uuid_module.uuid4())
        self._name = collection_data.get("name")
        self._recipe_ids = collection_data.get("recipeIds", [])

        self._anylist = anylist_instance
        self._fields_to_update = []

    @property
    def identifier(self) -> str:
        """Get collection identifier."""
        return self._identifier

    @property
    def name(self) -> Optional[str]:
        """Get collection name."""
        return self._name

    @name.setter
    def name(self, value: str) -> None:
        """Set collection name."""
        self._name = value
        self._mark_field_for_update("name")

    @property
    def recipe_ids(self) -> ListType[str]:
        """Get list of recipe IDs."""
        return self._recipe_ids.copy()

    def _mark_field_for_update(self, field_name: str) -> None:
        """Mark a field as needing update."""
        if field_name not in self._fields_to_update:
            self._fields_to_update.append(field_name)

    async def add_recipe(self, recipe_id: str) -> None:
        """
        Add a recipe to this collection.

        Args:
            recipe_id: ID of the recipe to add
        """
        if recipe_id not in self._recipe_ids:
            self._recipe_ids.append(recipe_id)

            data = {
                "operation": "add-recipe-to-collection",
                "collectionId": self._identifier,
                "recipeId": recipe_id,
            }

            await self._anylist._make_authenticated_request(
                "POST", "/data/recipe-data/update", json=data
            )

    async def remove_recipe(self, recipe_id: str) -> None:
        """
        Remove a recipe from this collection.

        Args:
            recipe_id: ID of the recipe to remove
        """
        if recipe_id in self._recipe_ids:
            self._recipe_ids.remove(recipe_id)

            data = {
                "operation": "remove-recipe-from-collection",
                "collectionId": self._identifier,
                "recipeId": recipe_id,
            }

            await self._anylist._make_authenticated_request(
                "POST", "/data/recipe-data/update", json=data
            )

    async def save(self) -> "RecipeCollection":
        """
        Save collection changes to the server.

        Returns:
            This collection instance
        """
        if not self._fields_to_update:
            return self

        data = {"operation": "save-recipe-collection", "collection": self.to_dict()}

        await self._anylist._make_authenticated_request(
            "POST", "/data/recipe-data/update", json=data
        )

        self._fields_to_update.clear()
        return self

    async def delete(self) -> None:
        """Delete this recipe collection."""
        data = {
            "operation": "delete-recipe-collection",
            "collectionId": self._identifier,
        }

        await self._anylist._make_authenticated_request(
            "POST", "/data/recipe-data/update", json=data
        )

    def to_dict(self) -> Dict[str, Any]:
        """Convert collection to dictionary representation."""
        return {
            "identifier": self._identifier,
            "name": self._name,
            "recipeIds": self._recipe_ids,
        }

    def __repr__(self) -> str:
        """String representation of collection."""
        return (
            f"RecipeCollection(identifier='{self._identifier}', "
            f"name='{self._name}', recipes={len(self._recipe_ids)})"
        )

File: lib/recipe-collection.js
const FormData = require('form-data');
const Joi = require('joi');
const uuid = require('./uuid');

// Pydantic-like schema for RecipeCollection validation
const RecipeCollectionSchema = Joi.object({
	identifier: Joi.string().default(() => uuid()),
	timestamp: Joi.number().default(() => Date.now() / 1000),
	name: Joi.string().required(),
	recipeIds: Joi.array().items(Joi.string()).default([]),
	collectionSettings: Joi.any().default(null),
}).options({
	stripUnknown: true,
	abortEarly: false,
});

/**
 * RecipeCollection class.
 * @class
 *
 * @param {object} recipeCollection recipeCollection
 * @param {object} context context
 *
 * @property {string} identifier
 * @property {string} timestamp
 * @property {string} name
 * @property {string[]} recipeIds
 * @property {string[]} preparationSteps
 * @property {string[]} photoIds
 * @property {string} adCampaignId
 * @property {string[]} photoUrls
 * @property {double} scaleFactor
 * @property {int} rating
 * @property {string} creationTimestamp
 * @property {string} nutritionalInfo
 * @property {int} cookTime
 * @property {int} prepTime
 * @property {string} servings
 * @property {string} paprikaIdentifier
 */
class RecipeCollection {
	/**
   * @hideconstructor
   */
	constructor(recipeCollection, {client, protobuf, uid, recipeDataId}) {
		// Validate input using pydantic-like schema
		const {error, value} = RecipeCollectionSchema.validate(recipeCollection || {});

		if (error) {
			const errorMessage = error.details.map(detail => detail.message).join(', ');
			throw new Error(`RecipeCollection validation failed: ${errorMessage}`);
		}

		this._client = client;
		this.protobuf = protobuf;
		this.uid = uid;
		this.recipeDataId = recipeDataId;

		this.identifier = value.identifier;
		this.timestamp = value.timestamp;
		this.name = value.name;
		this.recipeIds = value.recipeIds;
		this.collectionSettings = value.collectionSettings || new this.protobuf.PBRecipeCollectionSettings();
	}

	_encode() {
		return new this.protobuf.PBRecipeCollection({
			identifier: this.identifier,
			timestamp: this.timestamp,
			name: this.name,
			recipeIds: this.recipeIds,
			collectionSettings: this.collectionSettings,
		});
	}

	/**
   * Perform a recipe operation.
   * @private
   * @param {string} handlerId - Handler ID for the operation
   * @returns {Promise} - Promise representing the operation result
   */
	async performOperation(handlerId) {
		const ops = new this.protobuf.PBRecipeOperationList();
		const op = new this.protobuf.PBRecipeOperation();
		op.setMetadata({
			operationId: uuid(),
			handlerId,
			userId: this.uid,
		});

		// May not need recipedataid
		op.setRecipeDataId(this.recipeDataId);
		op.setRecipeCollection(this._encode());
		ops.setOperations(op);

		const form = new FormData();
		form.append('operations', ops.toBuffer());

		await this._client.post('data/user-recipe-data/update', {
			body: form,
		});
	}

	/**
		  * Save local changes to recipe to AnyList's API.
		  * @return {Promise}
		  */
	async save() {
		await this.performOperation('new-recipe-collection');
	}

	/**
		  * Delete a recipe collection from AnyList.
		  * @return {Promise}
		  */
	async delete() {
		await this.performOperation('remove-recipe-collection');
	}

	/**
		  * Adds an existing recipe to an existing recipe-collection on AnyList.
		  * @return {Promise}
		  */
	async addRecipe(recipeId) {
		if (recipeId) {
			this.recipeIds.push(recipeId);
			await this.performOperation('add-recipes-to-collection');
		}
	}

	/**
		  * Remove existing recipe from an existing recipe-collection on AnyList.
		  * @return {Promise}
		  */
	async removeRecipe(recipeId) {
		const recipeIdPos = this.recipeIds.indexOf(recipeId);
		if (recipeIdPos > -1) {
			await this.performOperation('remove-recipes-from-collection');
			this.recipeIds.splice(recipeIdPos, 1);
		}
	}
}
module.exports = RecipeCollection;
